File: medium/450. Delete Node in a BST/450. Delete Node in a BST.go
package main

/**
 * Definition for a binary tree node.
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findMin(node TreeNode) TreeNode {
	for node.Left != nil {
		node = *node.Left
	}

	return node
}

func updateNode(node *TreeNode) *TreeNode {
	if node.Left == nil {
		return node.Right
	}
	if node.Right == nil {
		return node.Left
	}

	minValue := findMin(*node.Right)
	node.Val = minValue.Val
	node.Right = deleteNode(node.Right, minValue.Val)
	return node
}

func run(node *TreeNode, key int) (bool, *TreeNode) {
	if node == nil {
		return false, nil
	}

	if node.Val == key {
		return true, updateNode(node)
	}

	status, right := run(node.Right, key)
	if status {
		node.Right = right
		return true, node
	} else {
		_, left := run(node.Left, key)
		node.Left = left
		return false, node
	}
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	_, res := run(root, key)
	return res
}

File: medium/450. Delete Node in a BST/450. Delete Node in a BST.py
from typing import Optional, Tuple


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findMin(self, node: TreeNode) -> TreeNode:
        while node.left:
            node = node.left
        return node

    def updateNode(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root.left:
            return root.right
        if not root.right:
            return root.left

        min_node = self.findMin(root.right)
        root.val = min_node.val
        root.right = self.deleteNode(root.right, min_node.val)

        return root

    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:

        def run(lief: Optional[TreeNode], key: int) -> Tuple[bool, Optional[TreeNode]]:
            if not lief:
                return False, None

            if lief.val == key:
                return True, self.updateNode(lief)

            right = run(lief.right, key)

            if right[0]:
                lief.right = right[1]
                return True, lief
            else:
                lief.left = run(lief.left, key)[1]
                return False, lief

        return run(root, key)[1]
